File: product_excluding_index.js

// Brute Force Solution 
function getProductsExcludingIndex(intArr){
  let results = [];

  for (let indexA = 0; indexA < intArr.length; indexA++) {
    let product = 1;
    for (let indexB = 0; indexB < intArr.length; indexB++) {
      if(indexA !== indexB){
        let currentInt = intArr[indexB];
        product = currentInt * product;
      }
    }
    results.push(product);
  }

  return results;
}

//Efficient Solution
function getProductsBeforeIndex(intArr){
  let products = [];
  let currentProduct = 1;
  for (const int of intArr) {
    products.push(currentProduct);
    currentProduct = int * currentProduct;
  }
  return products
}

function efficientGetProducts(intArr){
  let products = [];
  let productsBeforeIndex = getProductsBeforeIndex(intArr);
  let currentProduct = 1;
  if(intArr.length == 1) return products;
    for (let index = intArr.length - 1 ; index >= 0; index--) {
      products[index] = currentProduct * productsBeforeIndex[index];
      currentProduct = currentProduct * intArr[index];
    }
  return products;
}

File: solutions/product_excluding_index.rb
# Brute Force Solution - O(n^2)
def get_products_no_index(int_arr)
  results = []
  int_arr.each_with_index do |int_a, index_a|
    product = 1
    int_arr.each_with_index do |int_b, index_b|
      next if index_b == index_a
      product *= int_b
    end
    results.push(product)
  end
  results
end


# Most effcient solution - O(n)
# Greedy approach

def get_products_before_index(int_array)
  product_so_far = 1
  all_products_before = []
  for index in (0...int_array.length)
    #product before the first element in array is 1
    #therefore we don't want to calculate the produce
    #until we have pushed the product into the array
    all_products_before[index] = product_so_far
    product_so_far *= int_array[index]
  end
  all_products_before
end

def efficient_products_no_index(int_array)
  return [] if int_array.length == 1
  # anticipates edge case of array of length 1
  results = []
  products_before_index = get_products_before_index(int_array)
  # get array of products before index, push them into an array
  product_so_far = 1
  index = int_array.length - 1
  while index >= 0
    # iterate through product array backwards to calculate
    # the products except for the index
    results[index] = product_so_far * products_before_index[index]
    product_so_far *= int_array[index]
    index -= 1
  end
  results
end

